lib/internal/modules/esm/module_job.js

```diff
@@ -2,26 +2,80 @@
 
 const {
   ArrayPrototypeJoin,
+  Number,
   ObjectSetPrototypeOf,
   PromiseAll,
   SafeSet,
   SafePromise,
   StringPrototypeIncludes,
   StringPrototypeMatch,
-  StringPrototypeReplace,
   StringPrototypeSplit,
 } = primordials;
 
 const { ModuleWrap } = internalBinding('module_wrap');
 
 const { decorateErrorStack } = require('internal/util');
+const { fileURLToPath } = require('url');
 const assert = require('internal/assert');
 const resolvedPromise = SafePromise.resolve();
 
 function noop() {}
 
 let hasPausedEntry = false;
 
+function extractExample(file, lineNumber) {
+  const { readFileSync } = require('fs');
+  const { parse } = require('internal/deps/acorn/acorn/dist/acorn');
+  const { findNodeAt } = require('internal/deps/acorn/acorn-walk/dist/walk');
+
+  const code = readFileSync(file, { encoding: 'utf8' });
+  const parsed = parse(code, {
+    sourceType: 'module',
+    locations: true,
+  });
+  let start = 0;
+  let node;
+  do {
+    node = findNodeAt(parsed, start);
+    start = node.node.end + 1;
+
+    if (node.node.loc.end.line < lineNumber) {
+      continue;
+    }
+
+    if (node.node.type !== 'ImportDeclaration') {
+      continue;
+    }
+
+    const defaultSpecifier = node.node.specifiers.find(
+      (specifier) => specifier.type === 'ImportDefaultSpecifier'
+    );
+    const defaultImport = defaultSpecifier ?
+      defaultSpecifier.local.name :
+      'pkg';
+
+    let len = 0;
+    const imports = node.node.specifiers
+      .filter((specifier) => specifier.type === 'ImportSpecifier')
+      .map((specifier) => {
+        const statement =
+          specifier.local.name === specifier.imported.name ?
+            `${specifier.imported.name}` :
+            `${specifier.imported.name}: ${specifier.local.name}`;
+        len += statement.length;
+        return statement;
+      });
+
+    const destructuringAssignment = len > 80 - 17 ?
+      `\n${imports.map((i) => `  ${i}`).join(',\n')}\n` :
+      ` ${imports.join(', ')} `;
+
+    return `\n\nimport ${defaultImport} from '${node.node.source.value}';\n` +
+      `const {${destructuringAssignment}} = ${defaultImport};\n`;
+  } while (node === undefined || node.node.loc.start.line <= lineNumber);
+  return '';
```

Comment:
Can we imagine a case where the above loop would not find a matching import declaration?

The `return '';` was a desperate guard against having `undefined` appear in the error message in case the above loop does not find an import declaration… Any suggestions?

Reply:
What I wanted to say: I haven’t been able to reproduce this code path, so I should probably rather remove this speculative code. But maybe somebody has an idea if/how this path could be reproduced?

Reply:
Maybe you could throw a [`ERR_INTERNAL_ASSERTION`](https://nodejs.org/docs/latest/api/errors.html#ERR_INTERNAL_ASSERTION) instead? In this case, if someone stumbles upon it, they'll open an issue and it can be fixed.

Reply:
What's the best way to throw such an exception? Is there any precedent in the code base?

Reply:
You can check out `lib/internal/assert.js`

```js
const { ERR_INTERNAL_ASSERTION } = require('internal/errors').codes;
throw new ERR_INTERNAL_ASSERTION();
```

Reply:
I found `assert.fail()` being used in other internal js modules.